File: app/src/main/java/io/github/shadow578/music_dl/ui/more/MoreViewModel.java
package io.github.shadow578.music_dl.ui.more;

import android.app.Activity;
import android.app.Application;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.documentfile.provider.DocumentFile;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.mikepenz.aboutlibraries.LibsBuilder;

import java.util.Optional;
import java.util.concurrent.atomic.AtomicBoolean;

import io.github.shadow578.music_dl.LocaleOverride;
import io.github.shadow578.music_dl.R;
import io.github.shadow578.music_dl.backup.BackupData;
import io.github.shadow578.music_dl.backup.BackupHelper;
import io.github.shadow578.music_dl.downloader.TrackDownloadFormat;
import io.github.shadow578.music_dl.ui.base.BaseActivity;
import io.github.shadow578.music_dl.util.Async;
import io.github.shadow578.music_dl.util.preferences.Prefs;

/**
 * view model for more fragment
 */
public class MoreViewModel extends AndroidViewModel {
    public MoreViewModel(@NonNull Application application) {
        super(application);
    }

    /**
     * currently selected download format
     */
    private final MutableLiveData<TrackDownloadFormat> downloadFormat = new MutableLiveData<>(Prefs.DownloadFormat.get());

    /**
     * current state of ssl_fix enable
     */
    private final MutableLiveData<Boolean> enableSSLFix = new MutableLiveData<>(Prefs.EnableSSLFix.get());

    /**
     * current state of metadata tagging enable
     */
    private final MutableLiveData<Boolean> enableTagging = new MutableLiveData<>(Prefs.EnableMetadataTagging.get());

    /**
     * currently selected locale override
     */
    private final MutableLiveData<LocaleOverride> localeOverride = new MutableLiveData<>(Prefs.LocaleOverride.get());

    /**
     * open the about page
     *
     * @param activity parent activity
     */
    public void openAboutPage(@NonNull Activity activity) {
        final LibsBuilder libs = new LibsBuilder();
       // libs.setAboutShowIcon(true);
        libs.setAboutAppName(activity.getString(R.string.app_name));
        //libs.setShowLicense(true);
        //libs.setShowVersion(true);
        libs.start(activity);
    }

    /**
     * choose the download directory
     *
     * @param activity parent activity
     */
    public void chooseDownloadsDir(@NonNull Activity activity) {
        if (activity instanceof BaseActivity) {
            ((BaseActivity) activity).maybeSelectDownloadsDir(true);
        }
    }

    /**
     * import tracks from a backup file
     *
     * @param file the file to import from
     */
    public void importTracks(@NonNull DocumentFile file, @NonNull Activity parent) {
        Async.runAsync(() -> {
            // read the backup data
            final Optional<BackupData> backup = BackupHelper.readBackupData(getApplication(), file);
            if (!backup.isPresent()) {
                Async.runOnMain(()
                        -> Toast.makeText(getApplication(), R.string.restore_toast_failed, Toast.LENGTH_SHORT).show());
                return;
            }

            // show confirmation dialog
            Async.runOnMain(() -> {
                final AtomicBoolean replaceExisting = new AtomicBoolean(false);
                final int tracksCount = backup.get().tracks.size();
                new AlertDialog.Builder(parent)
                        .setTitle(getApplication().getString(R.string.restore_dialog_title, tracksCount))
                        .setSingleChoiceItems(R.array.restore_dialog_modes, 0, (dialog, mode)
                                -> replaceExisting.set(mode == 1))
                        .setNegativeButton(R.string.restore_dialog_negative, (dialog, w) -> dialog.dismiss())
                        .setPositiveButton(R.string.restore_dialog_positive, (dialog, w) -> {
                            // restore the backup
                            Toast.makeText(getApplication(), getApplication().getString(R.string.restore_toast_success, tracksCount), Toast.LENGTH_SHORT).show();
                            Async.runAsync(()
                                    -> BackupHelper.restoreBackup(getApplication(), backup.get(), replaceExisting.get()));
                        })
                        .show();
            });
        });
    }

    /**
     * export tracks to a backup file
     *
     * @param file the file to export to
     */
    public void exportTracks(@NonNull DocumentFile file) {
        Async.runAsync(() -> {
            final boolean success = BackupHelper.createBackup(getApplication(), file);
            if (!success) {
                Async.runOnMain(()
                        -> Toast.makeText(getApplication(), R.string.backup_toast_failed, Toast.LENGTH_SHORT).show());
            }
        });
    }

    /**
     * set the download format
     *
     * @param format new download format
     */
    public void setDownloadFormat(@NonNull TrackDownloadFormat format) {
        // ignore if no change
        if (format.equals(downloadFormat.getValue())) {
            return;
        }

        Prefs.DownloadFormat.set(format);
        downloadFormat.setValue(format);
    }

    /**
     * @return currently selected download format
     */
    @NonNull
    public LiveData<TrackDownloadFormat> getDownloadFormat() {
        return downloadFormat;
    }

    /**
     * set ssl fix enable
     *
     * @param enable enable ssl fix?
     */
    public void setEnableSSLFix(boolean enable) {
        if(Boolean.valueOf(enable).equals(enableSSLFix.getValue()))
        {
            return;
        }

        Prefs.EnableSSLFix.set(enable);
        enableSSLFix.setValue(enable);
    }

    /**
     * @return current state of ssl_fix enable
     */
    @NonNull
    public LiveData<Boolean> getEnableSSLFix() {
        return enableSSLFix;
    }

    /**
     * enable or disable metadata tagging
     *
     * @param enable is tagging enabled?
     */
    public void setEnableTagging(boolean enable) {
        if(Boolean.valueOf(enable).equals(enableTagging.getValue()))
        {
            return;
        }

        Prefs.EnableMetadataTagging.set(enable);
        enableTagging.setValue(enable);
    }

    /**
     * @return current state of metadata tagging enable
     */
    @NonNull
    public LiveData<Boolean> getEnableTagging() {
        return enableTagging;
    }

    /**
     * set the currently selected locale override
     *
     * @param localeOverride the currently selected locale override
     * @return  was the locale changed?
     */
    public boolean setLocaleOverride(@NonNull LocaleOverride localeOverride) {
        if(localeOverride.equals(this.localeOverride.getValue()))
        {
            return false;
        }

        Prefs.LocaleOverride.set(localeOverride);
        this.localeOverride.setValue(localeOverride);
        return true;
    }

    /**
     * @return the currently selected locale override
     */
    @NonNull
    public LiveData<LocaleOverride> getLocaleOverride() {
        return localeOverride;
    }
}

File: app/src/main/kotlin/io/github/shadow578/yodel/ui/more/MoreViewModel.kt
package io.github.shadow578.yodel.ui.more

import android.app.*
import android.content.Intent
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.documentfile.provider.DocumentFile
import androidx.lifecycle.*
import com.mikepenz.aboutlibraries.LibsBuilder
import io.github.shadow578.yodel.*
import io.github.shadow578.yodel.backup.BackupHelper
import io.github.shadow578.yodel.db.model.TrackStatus
import io.github.shadow578.yodel.downloader.DownloaderService
import io.github.shadow578.yodel.downloader.TrackDownloadFormat
import io.github.shadow578.yodel.ui.base.BaseActivity
import io.github.shadow578.yodel.ui.dev.DeveloperToolsActivity
import io.github.shadow578.yodel.util.*
import io.github.shadow578.yodel.util.preferences.Prefs
import java.util.concurrent.atomic.AtomicBoolean

/**
 * view model for more fragment
 */
class MoreViewModel(application: Application) : AndroidViewModel(application) {
    /**
     * currently selected download format
     */
    val downloadFormat = MutableLiveData(Prefs.DownloadFormat.get())

    /**
     * binder for [Prefs.EnableMetadataTagging]
     */
    val enableMetadataTaggingBinder = SwitchPreferenceBinder(Prefs.EnableMetadataTagging)

    /**
     * currently selected locale override
     */
    val localeOverride = MutableLiveData(Prefs.AppLocaleOverride.get())

    /**
     * how often [countAndOpenDeveloperTools] was called (== how often the app_icon was clicked)
     */
    private var developerToolsCounter: Int = 0

    /**
     * count how many times this function was called.
     * if it was called more than 5 times, open the developer tools activity
     *
     * @param parent parent activity
     */
    fun countAndOpenDeveloperTools(parent: Activity) {
        developerToolsCounter++
        if (developerToolsCounter >= 5) {
            parent.startActivity(Intent(parent, DeveloperToolsActivity::class.java))
            developerToolsCounter = 0
        }
    }

    /**
     * open the about page
     *
     * @param activity parent activity
     */
    fun openAboutPage(activity: Activity) {
        val libs = LibsBuilder()
        libs.aboutAppName = activity.getString(R.string.app_name)
        libs.start(activity)
    }

    /**
     * choose the download directory
     *
     * @param activity parent activity
     */
    fun chooseDownloadsDir(activity: Activity) {
        if (activity is BaseActivity)
            activity.maybeSelectDownloadsDir(true)
    }

    /**
     * import tracks from a backup file
     *
     * @param file the file to import from
     */
    fun importTracks(file: DocumentFile, parent: Activity) {
        val application = getApplication<Application>()
        val backupHelper = BackupHelper(getApplication())
        launchIO {
            // read the backup data
            val backup = backupHelper.readBackup(file)
            if (backup == null) {
                launchMain {
                    application.toast(
                            R.string.restore_toast_failed,
                            Toast.LENGTH_SHORT
                    )
                }
                return@launchIO
            }

            // show confirmation dialog
            launchMain {
                val replaceExisting = AtomicBoolean(false)
                val tracksCount = backup.tracks.size
                AlertDialog.Builder(parent)
                        .setTitle(
                                application.getString(
                                        R.string.restore_dialog_title,
                                        tracksCount
                                )
                        )
                        .setSingleChoiceItems(
                                R.array.restore_dialog_modes,
                                0
                        ) { _, mode -> replaceExisting.set(mode == 1) }
                        .setNegativeButton(R.string.restore_dialog_negative) { dialog, _ -> dialog.dismiss() }
                        .setPositiveButton(R.string.restore_dialog_positive) { _, _ ->
                            // restore the backup
                            application.toast(
                                    getApplication<Application>().getString(
                                            R.string.restore_toast_success,
                                            tracksCount
                                    ),
                                    Toast.LENGTH_SHORT
                            )

                            launchIO {
                                // restore, transform all restored to pending state
                                backupHelper.restoreBackup(
                                        backup,
                                        replaceExisting.get()
                                ) { status = TrackStatus.DownloadPending }

                                // start service on demand
                                DownloaderService.startOnDemand(application)
                            }
                        }
                        .show()
            }
        }
    }

    /**
     * export tracks to a backup file
     *
     * @param file the file to export to
     */
    fun exportTracks(file: DocumentFile) {
        launchIO {
            val success = BackupHelper(getApplication()).createBackup(file)
            if (!success) {
                launchMain {
                    getApplication<Application>().toast(
                            R.string.backup_toast_failed,
                            Toast.LENGTH_SHORT
                    )
                }
            }
        }
    }

    /**
     * set the download format
     *
     * @param format new download format
     */
    fun setDownloadFormat(format: TrackDownloadFormat) {
        // ignore if no change
        if (format == downloadFormat.value) {
            return
        }
        Prefs.DownloadFormat.set(format)
        downloadFormat.value = format
    }

    /**
     * set the currently selected locale override
     *
     * @param localeOverride the currently selected locale override
     * @return  was the locale changed?
     */
    fun setLocaleOverride(localeOverride: LocaleOverride): Boolean {
        if (localeOverride == this.localeOverride.value)
            return false

        Prefs.AppLocaleOverride.set(localeOverride)
        this.localeOverride.value = localeOverride
        return true
    }
}